### PR TITLE
feat(dsp): add error handling to controllers

### LINKED ISSUE #2796: dataspace protocol: add error support
# Feature Request

The protocol defines `CatalogError`, `NegotiationError`, and `TransferError`. These should be supported.

## Which Areas Would Be Affected?
dsp, api

## Why Is the Feature Desired?
dataspace protocol support

## Solution Proposal
Return in case of failure. Could be something that an `ExceptionMapper` could take care of.

### PR DESCRIPTION
**Pull request by juliapampus**
## What this PR changes/adds

Introduces `:dsp-core`. Implements error handling to dsp controllers following the spec.

## Why it does that

Proper error handling.

## Further notes

_List other areas of code that have changed but are not necessarily linked to the main feature. This could be method signature changes, package declarations, bugs that were encountered and were fixed inline, etc._

## Linked Issue(s)

Closes #2796
Closes #2917 

## Checklist

- [ ] added appropriate tests?
- [ ] performed checkstyle check locally?
- [ ] added/updated copyright headers?
- [ ] documented public classes/methods?
- [ ] added/updated relevant documentation?
- [ ] assigned appropriate label? (exclude from changelog with label `no-changelog`)
- [ ] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-edc/Connector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [Etiquette for pull requests](https://github.com/eclipse-edc/Connector/blob/main/pr_etiquette.md) for details_)

### CODE REVIEW

File: data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java [hunk 1]
```diff
@@ -75,43 +75,62 @@ public DspCatalogApiController(Monitor monitor, ObjectMapper mapper, IdentitySer
...
+        var claimToken = verificationResult.getContent();
 
-        var catalogJson = transformerRegistry.transform(catalog, JsonObject.class)
-                .orElseThrow(failure -> new EdcException(format("Failed to build response: %s", failure.getFailureDetail())));
+        Catalog catalog;
+        try {
+            catalog = service.getCatalog(message, claimToken).getContent();
+        } catch (Exception exception) {
+            var errorCode = UUID.randomUUID();
+            monitor.warning(String.format("DSP negotiation error id %s: %s", errorCode, exception));
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java:116]: This should state that it is an error returning the catalog, not a negotiation

File: data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java [hunk 2]
```diff
@@ -75,43 +75,62 @@ public DspCatalogApiController(Monitor monitor, ObjectMapper mapper, IdentitySer
...
+            return errorResponse(Response.Status.INTERNAL_SERVER_ERROR, String.format("Error code %s", errorCode));
+        }
 
-        var compacted = jsonLdService.compact(catalogJson);
-        if (compacted.failed()) {
-            throw new InvalidRequestException(compacted.getFailureDetail());
+        var catalogJson = transformerRegistry.transform(catalog, JsonObject.class);
+        if (catalogJson.failed()) {
+            var errorCode = UUID.randomUUID();
+            monitor.warning(String.format("DSP negotiation error id %s: %s", errorCode, catalogJson.getFailureMessages()));
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java:123]: Ditto

File: data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java [hunk 3]
```diff
@@ -75,43 +75,62 @@ public DspCatalogApiController(Monitor monitor, ObjectMapper mapper, IdentitySer
...
         }
-        //noinspection unchecked
-        return mapper.convertValue(compacted.getContent(), Map.class);
+
+        return Response.status(Response.Status.OK).type(MediaType.APPLICATION_JSON).entity(catalogJson.getContent()).build();
+    }
+
+    private Response errorResponse(Response.Status code, String message) {
+        return Response.status(code).type(MediaType.APPLICATION_JSON)
+                .entity(DspError.create(DSPACE_CATALOG_ERROR, Optional.empty(), String.valueOf(code.getStatusCode()), message))
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java:132]: DspError should be a builder. This will remove the need to pass Optional and make the interface fluent for adding problem messages.

Also, even though, in this case, code cannot be null, please don't use String.valueOf() because I came across a number of areas where it was in the codebase, which lead to difficult-to-find error because it converts null to "null."  toString() should be sufficient

File: data-protocols/dsp/dsp-core/src/main/java/org/eclipse/edc/protocol/dsp/DspError.java [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.Optional;
+import java.util.stream.Stream;
+
+import static org.eclipse.edc.protocol.dsp.type.DspPropertyAndTypeNames.DSPACE_PROCESS_ID;
+import static org.eclipse.edc.protocol.dsp.type.DspPropertyAndTypeNames.DSPACE_PROPERTY_CODE;
+import static org.eclipse.edc.protocol.dsp.type.DspPropertyAndTypeNames.DSPACE_PROPERTY_REASON;
+
+public final class DspError {
+
+    public static JsonObject create(String type, Optional<String> processId, String code, String... messages) {
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-core/src/main/java/org/eclipse/edc/protocol/dsp/DspError.java:30]: Per my previous comment, this should use a builder. It will add parameter safety, remove the need for optional, and make passing messages fluent.

Error will also need to synthesize a @context and set @vocab to the EDC namsespace.

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-api/build.gradle.kts [hunk 5]
```diff
@@ -19,8 +19,10 @@ plugins {
 
 dependencies {
     api(project(":data-protocols:dsp:dsp-api-configuration"))
+    api(project(":data-protocols:dsp:dsp-core"))
+    api(project(":data-protocols:dsp:dsp-http-core"))
 
-    // TODO remove this dependency by moving type constants to another module
+    // TODO remove this dependency by moving type constants to another module https://github.com/eclipse-edc/Connector/issues/3014
     api(project(":data-protocols:dsp:dsp-negotiation:dsp-negotiation-transform"))
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-api/build.gradle.kts:26]: Can't this be removed now?

**Review by jimmarino — COMMENTED** [review]
This looks really good. Comments to discuss inline.

File: data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java (hunk 1 above)
Comment (juliapampus) [reply data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java:116]: oh, copy & paste, sry :D

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-api/build.gradle.kts (hunk 5 above)
Comment (juliapampus) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-api/build.gradle.kts:26]: Probably, will check

File: data-protocols/dsp/dsp-core/src/main/java/org/eclipse/edc/protocol/dsp/DspError.java (hunk 4 above)
Comment (juliapampus) [reply data-protocols/dsp/dsp-core/src/main/java/org/eclipse/edc/protocol/dsp/DspError.java:30]: I will create a builder.

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-api/build.gradle.kts (hunk 5 above)
Comment (juliapampus) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-api/build.gradle.kts:26]: unfortunately, the controllers still have dependencies on constants in the transformer module. I already created #3014

File: data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java [hunk 6]
```diff
@@ -75,43 +74,75 @@ public DspCatalogApiController(Monitor monitor, ObjectMapper mapper, IdentitySer
...
-        return mapper.convertValue(compacted.getContent(), Map.class);
+
+        return Response.status(Response.Status.OK).type(MediaType.APPLICATION_JSON).entity(compacted.getContent()).build();
+    }
+
+    private Response errorResponse(Response.Status code, String message) {
+        return Response.status(code).type(MediaType.APPLICATION_JSON)
+                .entity(DspError.Builder.newInstance()
+                        .type(DSPACE_CATALOG_ERROR)
+                        .code(String.valueOf(code.getStatusCode()))
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java:142]: Please no String.valueOf() :-)

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-api/build.gradle.kts (hunk 5 above)
Comment (jimmarino) [reply data-protocols/dsp/dsp-negotiation/dsp-negotiation-api/build.gradle.kts:26]: ok

File: data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java (hunk 6 above)
Comment (juliapampus) [reply data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java:142]: Lol, I was 100% sure I committed an update. Should be correct now.

File: data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java [hunk 7]
```diff
@@ -75,43 +74,75 @@ public DspCatalogApiController(Monitor monitor, ObjectMapper mapper, IdentitySer
...
-        var catalog = service.getCatalog(message, claimToken)
-                .orElseThrow(exceptionMapper(Catalog.class));
+        var claimToken = verificationResult.getContent();
 
-        var catalogJson = transformerRegistry.transform(catalog, JsonObject.class)
-                .orElseThrow(failure -> new EdcException(format("Failed to build response: %s", failure.getFailureDetail())));
+        var catalog = service.getCatalog(message, claimToken);
+        if (catalog.failed()) {
+            var errorCode = UUID.randomUUID();
+            monitor.warning(String.format("DSP catalog error id %s: %s", errorCode, catalog.getFailureMessages()));
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java:117]: Can this be more descriptive, .e.g "Error returning catalog"?

File: data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java [hunk 8]
```diff
@@ -75,43 +74,75 @@ public DspCatalogApiController(Monitor monitor, ObjectMapper mapper, IdentitySer
...
+            return errorResponse(Response.Status.INTERNAL_SERVER_ERROR, String.format("Error code %s", errorCode));
+        }
 
-        var compacted = jsonLdService.compact(catalogJson);
+        var catalogJson = transformerRegistry.transform(catalog.getContent(), JsonObject.class);
+        if (catalogJson.failed()) {
+            var errorCode = UUID.randomUUID();
+            monitor.warning(String.format("DSP catalog error id %s: %s", errorCode, catalogJson.getFailureMessages()));
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java:124]: Can this be more descriptive, .e.g "Error transforming catalog"

File: data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java [hunk 9]
```diff
@@ -75,43 +74,75 @@ public DspCatalogApiController(Monitor monitor, ObjectMapper mapper, IdentitySer
...
+            return errorResponse(Response.Status.INTERNAL_SERVER_ERROR, String.format("Error code %s", errorCode));
+        }
+
+        var compacted = jsonLdService.compact(catalogJson.getContent());
         if (compacted.failed()) {
-            throw new InvalidRequestException(compacted.getFailureDetail());
+            var errorCode = UUID.randomUUID();
+            monitor.warning(String.format("DSP catalog error id %s: %s", errorCode, catalogJson.getFailureMessages()));
```
Comment (jimmarino) [inline data-protocols/dsp/dsp-catalog/dsp-catalog-api/src/main/java/org/eclipse/edc/protocol/dsp/catalog/api/controller/DspCatalogApiController.java:131]: Can this be more descriptive, .e.g "Error compacting catalog"

**Review by jimmarino — APPROVED** [review]
LGTM, just three minor suggestions to improve the error message

File: data-protocols/dsp/dsp-core/src/main/java/org/eclipse/edc/protocol/dsp/DspError.java [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+ */
+@JsonDeserialize(builder = DspError.Builder.class)
+public class DspError {
+
+    private String type;
+    private String processId;
+    private String code;
+    private List<String> messages;
+
+    public JsonObject toJson() {
```
Comment (ndr-brt) [inline data-protocols/dsp/dsp-core/src/main/java/org/eclipse/edc/protocol/dsp/DspError.java:43]: is not a good practice to add serialization details into a value object, we should follow the transformer pattern. Could be a refactor to be addressed in another PR.

Comment (jimmarino) [reply data-protocols/dsp/dsp-core/src/main/java/org/eclipse/edc/protocol/dsp/DspError.java:43]: Agreed, actually we don't need to deserialize this class, it's only for egress. If we need to deserialize errors, we should use a transformer

Comment (ndr-brt) [reply data-protocols/dsp/dsp-core/src/main/java/org/eclipse/edc/protocol/dsp/DspError.java:43]: > Agreed, actually we don't need to deserialize this class, it's only for egress. If we need to deserialize errors, we should use a transformer

I guess in the future we would need to deserialize it as well, e.g. to understand the error cause (on the requestor side)

Comment (jimmarino) [reply data-protocols/dsp/dsp-core/src/main/java/org/eclipse/edc/protocol/dsp/DspError.java:43]: When that happens we need to use a transformer and follow the DSP protocol spec. We can't assume an EDC sent the response so there's no need to use the type to deserialize it.

Comment (ndr-brt) [reply data-protocols/dsp/dsp-core/src/main/java/org/eclipse/edc/protocol/dsp/DspError.java:43]: `DspError` is following the protocol spec already, isn't it? All the messages we are sending/receiving pass through a specific type and related transformers, I think this should be the same for the errors, our representation doesn't need to be 1:1 with the json message, it just needs to be able to serialize to/deserialize from compliant jsons

Comment (jimmarino) [reply data-protocols/dsp/dsp-core/src/main/java/org/eclipse/edc/protocol/dsp/DspError.java:43]: We shouldn't use DspError to deserialize from the wire. When deserializing an error, I'm also not sure we need a transformer, although we can look into it. The main issue is the DSP error messages are intentionally loosely typed and may contain open-ended error information. This is best represented as Map (or JsonObject), logged in serialized form, and then the code interpreted by the dispatcher. I'm not sure it makes much sense to transform it because it can only be converted into a Map, which it already is.  

I don't think we should use transformers when serializing error messages since this increases complexity significantly for something that can be handled with a simple return type. We would also need to establish what happens when an error transformation fails.

**Review by ndr-brt — APPROVED** [review]

**Comment by jimmarino** [comment]
@paullatzelsperger  can you review so we can merge?

File: data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-api/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/api/controller/DspTransferProcessApiController.java [hunk 11]
```diff
@@ -176,17 +222,32 @@ public void transferProcessCompletion(@PathParam("id") String id, JsonObject jso
...
-        handleMessage(MessageSpec.Builder.newInstance(TransferTerminationMessage.class)
-                .processId(id)
-                .expectedMessageType(DSPACE_TRANSFER_TERMINATION_TYPE)
-                .message(jsonObject)
-                .token(token)
-                .serviceCall(protocolService::notifyTerminated)
-                .build());
+    public Response transferProcessTermination(@PathParam("id") String id, JsonObject jsonObject, @HeaderParam(AUTHORIZATION) String token) {
+        try {
+            handleMessage(MessageSpec.Builder.newInstance(TransferTerminationMessage.class)
```
Comment (paullatzelsperger) [inline data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-api/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/api/controller/DspTransferProcessApiController.java:231]: can't `handleMessage` just return a `Result` (or suitable subclass)? That way, we could keep in line with our coding style (returning results), and it would likely be much more efficient than throwing and immediately catching exceptions. 

Distinction between Auth failure and Invalid request could be handled through different `Failure` objects.

File: data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-api/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/api/controller/DspTransferProcessApiController.java [hunk 12]
```diff
@@ -176,17 +222,32 @@ public void transferProcessCompletion(@PathParam("id") String id, JsonObject jso
...
+                    .serviceCall(protocolService::notifyTerminated)
+                    .build());
+        } catch (AuthenticationFailedException exception) {
+            monitor.debug(String.format(UNAUTHORIZED + "Requested process id %s: %s", id, exception.getMessages()));
+            return errorResponse(Optional.of(id), Response.Status.UNAUTHORIZED, UNAUTHORIZED);
+        } catch (InvalidRequestException exception) {
+            monitor.debug(String.format(BAD_REQUEST + "Requested process id %s: %s", id, exception.getMessages()));
+            return errorResponse(Optional.of(id), Response.Status.BAD_REQUEST, BAD_REQUEST);
+        } catch (Exception exception) {
+            var errorCode = UUID.randomUUID();
```
Comment (paullatzelsperger) [inline data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-api/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/api/controller/DspTransferProcessApiController.java:245]: that seems spurious: we should not catch just _any_ exception. As discussed, only errors on the egress path (e.g. during transformation, compaction,...) should actually get return as an "explicit" 50x. Unexpected exceptions should just get bubbled up the stacktrace and be logged.

File: data-protocols/dsp/dsp-negotiation/dsp-negotiation-api/src/main/java/org/eclipse/edc/protocol/dsp/negotiation/api/controller/DspNegotiationApiController.java [hunk 13]
```diff
@@ -107,40 +112,53 @@ public DspNegotiationApiController(String callbackAddress,
...
      */
     @POST
     @Path(INITIAL_CONTRACT_REQUEST)
-    public Map<String, Object> initiateNegotiation(@RequestBody(description = DSPACE_NEGOTIATION_CONTRACT_REQUEST_MESSAGE, required = true)
+    public Response initiateNegotiation(@RequestBody(description = DSPACE_NEGOTIATION_CONTRACT_REQUEST_MESSAGE, required = true)
                                                    JsonObject jsonObject,
                                                    @HeaderParam(AUTHORIZATION) String token) {
+        ContractNegotiation negotiation;
+        try {
+            negotiation = handleMessage(MessageSpec.Builder.newInstance(ContractRequestMessage.class)
```
Comment (paullatzelsperger) [inline data-protocols/dsp/dsp-negotiation/dsp-negotiation-api/src/main/java/org/eclipse/edc/protocol/dsp/negotiation/api/controller/DspNegotiationApiController.java:137]: please check my comments in `DspTransferProcessApiController`, they apply here as well

**Review by paullatzelsperger — CHANGES_REQUESTED** [review]
almost there.

File: data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-api/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/api/controller/DspTransferProcessApiController.java (hunk 11 above)
Comment (juliapampus) [reply data-protocols/dsp/dsp-transfer-process/dsp-transfer-process-api/src/main/java/org/eclipse/edc/protocol/dsp/transferprocess/api/controller/DspTransferProcessApiController.java:228]: Created #3017.

**Review by paullatzelsperger — APPROVED** [review]
further refactoring to be done in #3026
